[Site Isolation][macOS] Transform mouse coordinates for drag events when forwarding to an iframe process

nipunshukla21 opened the pull request:
#### 02911a58052f17a12a466f6b74b25682bde67370
<pre>
[Site Isolation][macOS] Transform mouse coordinates for drag events when forwarding to an iframe process
<a href="https://bugs.webkit.org/show_bug.cgi?id=264611">https://bugs.webkit.org/show_bug.cgi?id=264611</a>
<a href="https://rdar.apple.com/118244400">rdar://118244400</a>

Reviewed by Charlie Wolfe.

Instead of directly passing dragData.clientPosition() to the drag controller we first transform it to map to the remote frame&apos;s coordinate space, similar to how this is already handled for mouse and touch events. This logic is implemented for both DragController::dragEnteredOrUpdated() and EventHandler::dragSourceEndedAt() along with accompanying APITests which confirm coordinate transformation as LayoutTests do not currently support testing drag and drop in WK2.

* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::dragEnteredOrUpdated):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::dragSourceEndedAt):
* Source/WebCore/page/EventHandler.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, DragEvents)):
(TestWebKitAPI::(SiteIsolation, DragSourceEndedAtCoordinateTransformation)):

Canonical link: <a href="https://commits.webkit.org/298858@main">https://commits.webkit.org/298858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9acc70cd9ca9b3c7019aa258046ce2f376279ec6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122659 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67158 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118476 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36954 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44836 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88550 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42981 "Too many flaky failures: fast/canvas/offscreen-no-script-context-crash.html, imported/w3c/web-platform-tests/event-timing/contextmenu.html, imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative.html, imported/w3c/web-platform-tests/trusted-types/set-event-handlers-content-attributes.tentative.html, js/dfg-arguments-alias-one-block-overwrite-arguments.html, js/dom/random-array-gc-stress.html, js/dom/webidl-type-mapping.html, js/large-expressions.html, js/structuredClone/structured-clone-of-CachedString-in-map.html, js/structuredClone/structured-clone-of-ResizableSharedArrayBuffer.html, performance-api/performance-observer-exception.html, performance-api/performance-observer-nested.html, storage/indexeddb/cursor-overloads-private.html, webgl/2.0.0/conformance2/textures/canvas_sub_rectangle/tex-2d-rgba32f-rgba-float.html, webgl/2.0.0/conformance2/textures/image/tex-2d-rgba16f-rgba-half_float.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_blob/tex-2d-rgba16f-rgba-float.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_canvas/tex-2d-r16f-red-half_float.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119536 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104595 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69010 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28541 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22700 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66326 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98856 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22860 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125797 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43486 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32669 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97231 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43850 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100797 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97024 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24757 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42333 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20260 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39448 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43369 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48972 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42838 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46178 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44544 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -5445,4 +5445,177 @@ HTTPServer server({
+#if ENABLE(DRAG_SUPPORT) && PLATFORM(MAC)
+TEST(SiteIsolation, DragEnteredOrUpdatedCoordinateTransformation)
+{
+    auto mainframeHTML = "<script>"
```
nit: the HTML markups can be `static constexpr`

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -5445,4 +5445,177 @@ HTTPServer server({
+        { "/subframe"_s, { subframeHTML } }
+    }, HTTPServer::Protocol::HttpsProxy);
+
+    auto navigationDelegate = adoptNS([TestNavigationDelegate new]);
```
```suggestion
    RetainPtr navigationDelegate = adoptNS([TestNavigationDelegate new]);
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -5445,4 +5445,177 @@ HTTPServer server({
+    [navigationDelegate allowAnyTLSCertificate];
+    auto configuration = server.httpsProxyConfiguration();
+    enableSiteIsolation(configuration);
+    auto simulator = adoptNS([[DragAndDropSimulator alloc] initWithWebViewFrame:NSMakeRect(0, 0, 600, 600) configuration:configuration]);
```
```suggestion
    RetainPtr simulator = adoptNS([[DragAndDropSimulator alloc] initWithWebViewFrame:NSMakeRect(0, 0, 600, 600) configuration:configuration]);
```

aprotyas reviewed (commented):
some nit comments; they apply equally to both the API tests

charliewolfe commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4391,8 +4391,18 @@ std::optional<RemoteUserInputEventData> EventHandler::dragSourceEndedAt(const Pl
     if (RefPtr remoteSubframe = dynamicDowncast<RemoteFrame>(subframeForHitTestResult(mouseEvent))) {
-        // FIXME(264611): These mouse coordinates need to be correctly transformed.
-        return RemoteUserInputEventData { remoteSubframe->frameID(),  mouseEvent.hitTestResult().roundedPointInInnerNodeFrame() };
+        RefPtr frameView = m_frame->view();
```
Can't you just use `EventHandler::userInputEventDataForRemoteFrame` and avoid duplicating this code?

charliewolfe commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -323,8 +324,18 @@ Variant<std::optional<DragOperation>, RemoteUserInputEventData> DragController::
     if (RefPtr remoteSubframe = dynamicDowncast<RemoteFrame>(frame.eventHandler().subframeForTargetNode(hitTestResult.targetNode()))) {
-        // FIXME(264611): These mouse coordinates need to be correctly transformed.
-        return RemoteUserInputEventData { remoteSubframe->frameID(), dragData.clientPosition() };
+        RefPtr frameView = frame.view();
```
Same thing here, you should be able to use `EventHandler::userInputEventDataForRemoteFrame`. You may need to make the function public though.

charliewolfe commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -5440,4 +5440,166 @@ HTTPServer server({
 }
 
+#if ENABLE(DRAG_SUPPORT) && PLATFORM(MAC)
+TEST(SiteIsolation, DragEnteredOrUpdatedCoordinateTransformation)
```
Adding these new tests is fine if needed, but we already have `SiteIsolation.DragEvents` which verifies that the drag events fire in a site isolation iframe. You could probably just add the coordinates to the `postMessage` in that test.

nipunshukla21 replied:
I moved most of this logic to `DragEvents` and deleted the `DragEnteredOrUpdatedCoordinateTransformation` test, however I kept `DragSourceEndedAtCoordinateTransformation` as `DragEvents` also looks for a `dragleave` which conflicts with the goal of exercising the transformation for `dragend` within a remote frame, which the latter test takes care of.

charliewolfe commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4392,8 +4392,8 @@ std::optional<RemoteUserInputEventData> EventHandler::dragSourceEndedAt(const Pl
     if (RefPtr remoteSubframe = dynamicDowncast<RemoteFrame>(subframeForHitTestResult(mouseEvent))) {
-        // FIXME(264611): These mouse coordinates need to be correctly transformed.
-        return RemoteUserInputEventData { remoteSubframe->frameID(),  mouseEvent.hitTestResult().roundedPointInInnerNodeFrame() };
+        IntPoint pointInFrame = mouseEvent.hitTestResult().roundedPointInInnerNodeFrame();
```
nit - auto
```suggestion
        auto pointInFrame = mouseEvent.hitTestResult().roundedPointInInnerNodeFrame();
```

charliewolfe commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -323,8 +323,10 @@ Variant<std::optional<DragOperation>, RemoteUserInputEventData> DragController::
     if (RefPtr remoteSubframe = dynamicDowncast<RemoteFrame>(frame.eventHandler().subframeForTargetNode(hitTestResult.targetNode()))) {
-        // FIXME(264611): These mouse coordinates need to be correctly transformed.
-        return RemoteUserInputEventData { remoteSubframe->frameID(), dragData.clientPosition() };
+        IntPoint pointInFrame = hitTestResult.roundedPointInInnerNodeFrame();
```
nit - auto
```suggestion
        auto pointInFrame = hitTestResult.roundedPointInInnerNodeFrame();
```

charliewolfe commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1836,11 +1837,46 @@ HTTPServer server({
+    if (dragenterComponents.count == 2) {
+        int x = [dragenterComponents[0] intValue];
+        int y = [dragenterComponents[1] intValue];
+        EXPECT_TRUE(x >= 65 && x <= 75) << "Expected dragenter x coordinate around 71, got " << x;
```
Do we need this because the results aren't consistent?

nipunshukla21 replied:
They seem very consistent on my end, but it seems like the drag simulator polls at some unknown (to me) interval. When I added the range I was thinking more of the dragover events vs dragenter but my thought process was that I don't want some small change in the drag simulator to break this test, and even with the range we're still exercising the transformation functionality.

nipunshukla21 committed:
[Site Isolation][macOS] Transform mouse coordinates for drag events when forwarding to an iframe process
https://bugs.webkit.org/show_bug.cgi?id=264611
rdar://118244400

Reviewed by Charlie Wolfe.

Instead of directly passing dragData.clientPosition() to the drag controller we first transform it to map to the remote frame's coordinate space, similar to how this is already handled for mouse and touch events. This logic is implemented for both DragController::dragEnteredOrUpdated() and EventHandler::dragSourceEndedAt() along with accompanying APITests which confirm coordinate transformation as LayoutTests do not currently support testing drag and drop in WK2.

* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::dragEnteredOrUpdated):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::dragSourceEndedAt):
* Source/WebCore/page/EventHandler.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, DragEvents)):
(TestWebKitAPI::(SiteIsolation, DragSourceEndedAtCoordinateTransformation)):

Canonical link: https://commits.webkit.org/298858@main
